proto/keystore_api/v1/keystore.proto

```diff
@@ -0,0 +1,106 @@
+// Message content encoding structures
+syntax = "proto3";
+
+package xmtp.keystore_api.v1;
+
+import "message_contents/public_key.proto";
+import "message_contents/invitation.proto";
+import "message_contents/ciphertext.proto";
+import "message_api/v1/message_api.proto";
+
+enum ErrorCode {
+    ERROR_CODE_UNSPECIFIED = 0;
+    ERROR_CODE_INVALID_INPUT = 1;
+    ERROR_CODE_NO_MATCHING_PREKEY = 2;
+}
+
+message KeystoreError {
+    string message = 1;
+    ErrorCode code = 2;
+}
+
+message ConversationReference {
+    string topic = 1;
+    uint64 createdAtNs = 2;
+    xmtp.message_contents.InvitationV1.Context context = 3;
+}
+
+message DecryptV1Request {
+    message Request {
+        xmtp.message_contents.Ciphertext payload = 1;
+        xmtp.message_contents.PublicKeyBundle peerKeys = 2;
+        bytes header_bytes = 3;
+        bool is_sender = 4;
+    }
+
+    repeated Request requests = 1;
+}
+
+message DecryptResponse {
+    oneof response {
+        bytes decrypted = 1;
+        KeystoreError error = 2;
+    }
+}
+
+message DecryptV2Request {
+    message Request {
+        xmtp.message_contents.Ciphertext payload = 1;
+        bytes header_bytes = 2;
+        string content_topic = 3;
+    }
+
+    repeated Request requests = 1;
+}
+
+message EncryptV1Request {
+    message Request {
+        xmtp.message_contents.PublicKeyBundle recipient = 1;
+        bytes payload = 2;
+        bytes header_bytes = 3;
+    }
+
+    repeated Request requests = 1;
+}
+
+message EncryptResponse {
+    oneof response {
+        xmtp.message_contents.Ciphertext encrypted = 1;
+        KeystoreError error = 2;
+    }
+}
+
+message EncryptV2Request {
+    message Request {
+        bytes payload = 1;
+        bytes header_bytes = 2;
+        string content_topic = 3;
+    }
+
+    repeated Request requests = 1;
+}
+
+message CreateInviteRequest {
+    xmtp.message_contents.InvitationV1.Context context = 1;
+    xmtp.message_contents.SignedPublicKeyBundle recipient = 2;
+    uint64 createdAtNs = 3;
+}
+
+message CreateInviteResponse {
```

Comment:
I decided the right heuristic here is to have batch responses return errors in the response array and single-request endpoints simply throw.